File: 907/solution.go
const mod = 1e9 + 7

type Pair struct {
    value, weight interface{}
}

func sumSubarrayMins(arr []int) int {
    stk := make([] Pair, 0)
    
    answer, sum := 0, 0
    for _, val := range(arr) {
        weight := 1
        for len(stk) > 0 && val < stk[len(stk) - 1].value.(int) {
            sum -= stk[len(stk) - 1].value.(int)*stk[len(stk) - 1].weight.(int)
            weight += stk[len(stk) - 1].weight.(int)
            stk = stk[:len(stk) - 1]
        }
        stk = append(stk, Pair{val, weight})
        sum += val*weight
        answer = (answer + sum % mod) % mod
    }
    return answer
}

File: 907/solution.cpp
class Solution {
public:
    const int mod = 1e9 + 7;
    int sumSubarrayMins(vector<int>& arr) {
        stack<pair<int, int>> stk; // (value, accumulated_weight)
        
        int answer = 0;
        long long sum = 0;
        for (int val: arr) {
            int weight = 1;
            while (!stk.empty() && val < stk.top().first) {
                sum -= stk.top().first*stk.top().second;
                weight += stk.top().second;
                stk.pop();
            }
            stk.push({ val, weight });
            sum += val*weight;
            answer = (answer + sum % mod) % mod;
        }
        return answer;
    }
};